### DISCUSSION TITLE
How to define a daemon thread of type IO Void?

### DISCUSSION THREAD

**Discussion by istathar** [awkward-squad/ki#15] [General]
I'm sure I'm missing something obvious but for a thread that "never returns" `fork_` needs an IO Void.

How do you make an IO Void? I mean I know it could be `undefined` which is of type a but that seems suboptimal!

The only example I could find is in your test suite and you're using `throwIO` (which is also of type a, of course). Is that what you had in mind, that every daemon thread that "never returns" would end with a `throw`?

**Comment by evanrelf** [comment]
You can use [`forever`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Monad.html#v:forever):

```haskell
forever :: Applicative f => f a -> f b
```

For example:

```haskell
echo :: IO Void
echo = forever do
  input <- getLine
  putStrLn input

**Comment by ocharles** [comment]
More generally, anything that won't terminate will be a computation of type `IO a`, which fits the requirement for `IO Void`

**Comment by mitchellwrosen** [comment]
Yep, ditto to both of the previous comments. You could use `forever` as @evanrelf suggests, or if your daemon manually calls itself, either way - just ascribe it either the type `IO a` or `IO Void`.

**Comment by evanrelf** [comment]
I think of it like this:

A function's return type is a promise that you'll return that thing once you're done running.

But if you never finish running, you never have to make good on that promise, so you can make wild claims.

It's impossible to return `Void`, but you can claim you will once you're done, because you'll never be done 🙂
